File: mdw-admin/web/js/ui/step.js
// Copyright (c) 2016 CenturyLink, Inc. All Rights Reserved.
'use strict';

var stepMod = angular.module('mdwStep', ['mdw']);

stepMod.factory('Step', ['mdw', 'util', 'DC',
                         function(mdw, util, DC) {
  
  var Step = function(activity) {
    this.activity = activity;
    this.workflowType = 'activity';
  };
  
  Step.INST_W = 8;
  Step.OLD_INST_W = 4;
  Step.MAX_INSTS = 10;
  
  Step.STATUSES = [
    {status: 'Unknown', color: 'transparent'},
    {status: 'Pending', color: 'blue'},
    {status: 'In Progress', color: 'green'},
    {status: 'Failed', color: 'red'},
    {status: 'Completed', color: 'black'},
    {status: 'Canceled', color: 'darkgray'},
    {status: 'Hold', color: 'cyan'},
    {status: 'Waiting', color: 'yellow'}
  ];

  Step.prototype.draw = function(diagram) {
    var activity = this.workflowObj = this.activity;
    var shape;
    if (this.implementor.icon && this.implementor.icon.startsWith('shape:'))
      shape = this.implementor.icon.substring(6);
    
    // runtime state first
    if (this.instances) {
      var adj = 0;
      if (shape == 'start' || shape == 'stop')
        adj = 2;
      diagram.drawState(this.display, this.instances, !diagram.drawBoxes, adj);
    }
    
    if (this.implementor.icon) {
      var yAdjust = -2;
      if (shape) {
        if ('start' == shape) {
          diagram.drawOval(this.display.x, this.display.y, this.display.w, this.display.h, 'green', 'white');
        }
        else if ('stop' == shape) {
          diagram.drawOval(this.display.x, this.display.y, this.display.w, this.display.h, 'red', 'white');
        }
        else if ('decision' == shape) {
          diagram.drawDiamond(this.display.x, this.display.y, this.display.w, this.display.h);
          yAdjust = -8;
        }
        else if ('activity' == shape) {
          diagram.roundedRect(this.display.x, this.display.y, this.display.w, this.display.h, DC.BOX_OUTLINE_COLOR);
        }
      }
      else {
        if (diagram.drawBoxes)
          diagram.roundedRect(this.display.x, this.display.y, this.display.w, this.display.h, DC.BOX_OUTLINE_COLOR);
        var iconSrc = mdw.roots.hub + '/asset/' + this.implementor.icon;
        var iconX = this.display.x + this.display.w / 2 - 12;
        var iconY = this.display.y + 5;
        diagram.drawImage(iconSrc, iconX, iconY);
        yAdjust = +4; 
      }
    }
    else {
      diagram.roundedRect(this.display.x, this.display.y, this.display.w, this.display.h, DC.BOX_OUTLINE_COLOR);
    }

    // title
    this.title.lines.forEach(function(line) {
      diagram.context.fillText(line.text, line.x, line.y + yAdjust);
    });
    
    // logical id
    diagram.context.fillStyle = DC.META_COLOR;
    diagram.context.fillText(activity.id, this.display.x + 2, this.display.y - 2);
    diagram.context.fillStyle = DC.DEFAULT_COLOR;
    
  };
  
  // sets display/title and returns an object with w and h for required size
  Step.prototype.prepareDisplay = function(diagram) {
    var maxDisplay = { w: 0, h: 0};
    var display = diagram.getDisplay(this.activity.attributes.WORK_DISPLAY_INFO);

    if (display.x + display.w > maxDisplay.w)
      maxDisplay.w = display.x + display.w;
    if (display.y + display.h > maxDisplay.h)
      maxDisplay.h = display.y + display.h;
    
    // step title
    var titleLines = [];
    this.activity.name.getLines().forEach(function(line) {
      titleLines.push({ text: line });
    });
    var title = { text: this.activity.name, lines: titleLines, w: 0, h:0 };
    for (var i = 0; i < title.lines.length; i++) {
      var line = title.lines[i];
      var textMetrics = diagram.context.measureText(line.text);
      if (textMetrics.width > title.w)
        title.w = textMetrics.width;
      title.h += DC.DEFAULT_FONT_SIZE;
      line.x = display.x + display.w / 2 - textMetrics.width / 2;
      line.y = display.y + display.h / 2 + DC.DEFAULT_FONT_SIZE * (i + 0.5);
      if (line.x + textMetrics.width > maxDisplay.w)
        maxDisplay.w = line.x + textMetrics.width;
      if (line.y + DC.DEFAULT_FONT_SIZE > maxDisplay.h)
        maxDisplay.h = line.y + DC.DEFAULT_FONT_SIZE;
    }

    this.display = display;
    this.title = title;    
    
    return maxDisplay;
  };
  
  Step.prototype.applyState = function(activityInstances) {
    this.instances = activityInstances;
  };
      
  return Step;
    
}]);

File: mdw-common/src/com/centurylink/mdw/draw/Step.java
/**
 * Copyright (c) 2014 CenturyLink, Inc. All Rights Reserved.
 */
package com.centurylink.mdw.draw;

import java.awt.Color;
import java.awt.Graphics2D;

import com.centurylink.mdw.model.workflow.Activity;

public class Step extends Shape implements Drawable {
    private final Activity activity;
    private final Implementor implementor;
    private final boolean boxStyle;
    public static final int MIN_SIZE = 4;

    public Step(Graphics2D g2d, Activity activity, Implementor implementor, boolean boxStyle) {
        super(g2d, new Display(activity.getAttribute("WORK_DISPLAY_INFO")));
        this.g2d = g2d;
        this.activity = activity;
        this.implementor = implementor;
        this.boxStyle = boxStyle;
    }

    public final Activity getActivity() {
        return this.activity;
    }

    public Display draw() {
        Display extents = new Display(0, 0, this.getDisplay().getX() + this.getDisplay().getW(),
                this.getDisplay().getY() + this.getDisplay().getH());
        int yAdjust = -3;
        Color textColor = Display.DEFAULT_COLOR;
        int iconX;
        int w;
        int y;
        if (implementor.getIcon() != null) {
            if (boxStyle) {
                drawRect(this.getDisplay().getX(), this.getDisplay().getY(),
                        this.getDisplay().getW(), this.getDisplay().getH(), null, null, null);
            }
            iconX = display.getX() + display.getW() / 2 - 12;
            int iconY = display.getY() + 5;
            extents.setW(Math.max(extents.getW(), iconX + implementor.getIcon().getIconWidth()));
            extents.setH(Math.max(extents.getH(), iconY + implementor.getIcon().getIconHeight()));
            drawIcon(implementor.getIcon(), iconX, iconY, 1.0F);
        }
        else if (this.implementor.getIconName() != null) {
            String shape = this.implementor.getIconName().substring(6);
            if (shape.equals("stop")) {
                drawOval(this.getDisplay().getX(), this.getDisplay().getY(),
                        this.getDisplay().getW(), this.getDisplay().getH(), Display.STOP_COLOR,
                        null, g2d);
                textColor = Display.SHAPE_TEXT_COLOR;
            }
            else if (shape.equals("start")) {
                drawOval(this.getDisplay().getX(), this.getDisplay().getY(),
                        this.getDisplay().getW(), this.getDisplay().getH(), Display.START_COLOR,
                        null, g2d);
                textColor = Display.SHAPE_TEXT_COLOR;
            }
            else if (shape.equals("decision")) {
                drawDiamond(this.getDisplay().getX(), this.getDisplay().getY(),
                        this.getDisplay().getW(), this.getDisplay().getH(), null);
            }
            else
                drawRect(this.getDisplay().getX(), this.getDisplay().getY(),
                        this.getDisplay().getW(), this.getDisplay().getH(), null, null, null);
        }
        else
            drawRect(this.getDisplay().getX(), this.getDisplay().getY(), this.getDisplay().getW(),
                    this.getDisplay().getH(), null, null, null);

        if (this.activity.getName() != null) {
            String[] lines = this.activity.getName().split("\r\n");
            w = 0;
            y = this.getDisplay().getY() + this.getDisplay().getH() / 2;
            if (lines.length == 1) {
                y += g2d.getFontMetrics().getHeight() / 2;
            }

            if (this.implementor.getIcon() != null) {
                y += this.implementor.getIcon().getIconHeight() / 2;
            }

            if (y < 0) {
                y = 0;
            }

            for (String line : lines) {
                int lw = this.g2d.getFontMetrics().stringWidth(line);
                if (lw > w) {
                    w = lw;
                }

                int x = this.getDisplay().getX() + this.getDisplay().getW() / 2 - lw / 2;
                extents.setW(Math.max(extents.getW(), x + w));
                extents.setH(
                        Math.max(extents.getH(), y + yAdjust + g2d.getFontMetrics().getHeight()));
                drawText(line, x, y + yAdjust, null, textColor);
                y += g2d.getFontMetrics().getHeight() - 1;
            }
        }

        if (this.activity.getId() > 0L) {
            extents.setW(Math.max(extents.getW(), this.display.getX() + 2
                    + g2d.getFontMetrics().stringWidth(activity.getId().toString())));
            extents.setH(Math.max(extents.getH(),
                    this.display.getY() - 2 + g2d.getFontMetrics().getHeight()));
            drawText(activity.getLogicalId(), this.display.getX() + 2, this.display.getY() - 2,
                    null, Display.META_COLOR);
        }

        return extents;
    }
}
